File: Decyx/decyx/dialogs/prompt_review_dialog.py
# prompt_review_dialog.py

import threading
import javax.swing as swing
from javax.swing import JFrame, JPanel, JButton, JScrollPane, BoxLayout, JLabel, JTextArea
from java.awt import BorderLayout

from decyx.config import SKIP_PROMPT_CONFIRMATION

# Dialog Class
class PromptReviewDialog(JFrame):
    """
    Dialog for reviewing and editing the prompt before sending it to the Claude API.
    """
    def __init__(self, prompt, title, on_submit, on_cancel):
        super(PromptReviewDialog, self).__init__(title)
        self.prompt = prompt
        self.on_submit = on_submit
        self.on_cancel = on_cancel
        self.final_prompt = None
        self.init_ui()

    def init_ui(self):
        try:
            panel = JPanel()
            panel.setLayout(BoxLayout(panel, BoxLayout.Y_AXIS))

            self.add_instruction_label(panel)
            self.add_prompt_text_area(panel)
            self.add_buttons(panel)

            self.getContentPane().add(JScrollPane(panel), BorderLayout.CENTER)
            self.setSize(700, 500)
            self.setLocationRelativeTo(None)
            self.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE)
            self.setVisible(True)
        except Exception as e:
            print "Error initializing PromptReviewDialog UI: {}".format(e)

    def add_instruction_label(self, panel):
        try:
            instruction_label = JLabel("Review and edit the final prompt before sending to Claude API:")
            instruction_label.setToolTipText("You can modify the prompt as needed")
            panel.add(instruction_label)
        except Exception as e:
            print "Error adding instruction label: {}".format(e)

    def add_prompt_text_area(self, panel):
        try:
            self.prompt_text_area = JTextArea(self.prompt, 20, 60)
            self.prompt_text_area.setLineWrap(True)
            self.prompt_text_area.setWrapStyleWord(True)
            self.prompt_text_area.setToolTipText("Edit the prompt here")
            panel.add(JScrollPane(self.prompt_text_area))
        except Exception as e:
            print "Error adding prompt text area: {}".format(e)

    def add_buttons(self, panel):
        try:
            button_panel = JPanel()
            send_button = JButton("Send to Claude API")
            send_button.addActionListener(lambda e: self.send())
            send_button.setToolTipText("Send the prompt to the Claude API")
            cancel_button = JButton("Cancel")
            cancel_button.addActionListener(lambda e: self.cancel())
            cancel_button.setToolTipText("Cancel and discard changes")
            button_panel.add(send_button)
            button_panel.add(cancel_button)
            self.getContentPane().add(button_panel, BorderLayout.SOUTH)
        except Exception as e:
            print "Error adding buttons to PromptReviewDialog: {}".format(e)

    def send(self):
        try:
            self.final_prompt = self.prompt_text_area.getText()
            self.on_submit(self.final_prompt)
            self.dispose()
        except Exception as e:
            print "Error sending prompt: {}".format(e)

    def cancel(self):
        try:
            self.final_prompt = None
            self.on_cancel()
            self.dispose()
        except Exception as e:
            print "Error cancelling PromptReviewDialog: {}".format(e)

# Helper Functions
def show_prompt_review_dialog(prompt, title):
    """
    Displays the PromptReviewDialog and waits for user interaction.

    Args:
        prompt (str): The prompt text to review.
        title (str): The title of the dialog window.

    Returns:
        str or None: The final prompt after user edits or None if cancelled.
    """
    if SKIP_PROMPT_CONFIRMATION:
        return prompt

    final_prompt = []
    dialog_complete = threading.Event()

    def on_submit(final_p):
        final_prompt.append(final_p)
        dialog_complete.set()

    def on_cancel():
        final_prompt.append(None)
        dialog_complete.set()

    def create_dialog():
        PromptReviewDialog(prompt, title, on_submit, on_cancel)

    swing.SwingUtilities.invokeLater(create_dialog)
    dialog_complete.wait()
    return final_prompt[0]

File: Decyx/decyx/dialogs/PromptReviewDialog.java
package decyx.dialogs;

import decyx.DecyxConfig;

import javax.swing.*;
import java.awt.*;
import java.util.concurrent.CountDownLatch;

/**
 * Dialog for reviewing and editing the prompt before sending it to the Claude API.
 */
public class PromptReviewDialog extends JFrame {

    private String finalPrompt;
    private final Runnable onComplete;
    private JTextArea promptTextArea;

    public PromptReviewDialog(String prompt, String title, Runnable onComplete) {
        super(title);
        this.finalPrompt = null;
        this.onComplete = onComplete;
        initUI(prompt);
    }

    private void initUI(String prompt) {
        try {
            JPanel panel = new JPanel();
            panel.setLayout(new BoxLayout(panel, BoxLayout.Y_AXIS));

            addInstructionLabel(panel);
            addPromptTextArea(panel, prompt);
            addButtons(panel);

            getContentPane().add(new JScrollPane(panel), BorderLayout.CENTER);
            setSize(700, 500);
            setLocationRelativeTo(null);
            setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
            setVisible(true);
        } catch (Exception e) {
            System.out.println("Error initializing PromptReviewDialog UI: " + e.getMessage());
        }
    }

    private void addInstructionLabel(JPanel panel) {
        JLabel instructionLabel = new JLabel("Review and edit the final prompt before sending to Claude API:");
        instructionLabel.setToolTipText("You can modify the prompt as needed");
        panel.add(instructionLabel);
    }

    private void addPromptTextArea(JPanel panel, String prompt) {
        promptTextArea = new JTextArea(prompt, 20, 60);
        promptTextArea.setLineWrap(true);
        promptTextArea.setWrapStyleWord(true);
        promptTextArea.setToolTipText("Edit the prompt here");
        panel.add(new JScrollPane(promptTextArea));
    }

    private void addButtons(JPanel panel) {
        JPanel buttonPanel = new JPanel();
        JButton sendButton = new JButton("Send to Claude API");
        sendButton.addActionListener(e -> send());
        sendButton.setToolTipText("Send the prompt to the Claude API");
        JButton cancelButton = new JButton("Cancel");
        cancelButton.addActionListener(e -> cancel());
        cancelButton.setToolTipText("Cancel and discard changes");
        buttonPanel.add(sendButton);
        buttonPanel.add(cancelButton);
        getContentPane().add(buttonPanel, BorderLayout.SOUTH);
    }

    private void send() {
        finalPrompt = promptTextArea.getText();
        onComplete.run();
        dispose();
    }

    private void cancel() {
        finalPrompt = null;
        onComplete.run();
        dispose();
    }

    public String getFinalPrompt() {
        return finalPrompt;
    }

    /**
     * Displays the dialog and waits for user interaction.
     * If SKIP_PROMPT_CONFIRMATION is true, returns the prompt immediately.
     *
     * @param prompt the prompt text to review
     * @param title  the dialog title
     * @return the final prompt after user edits or null if cancelled
     */
    public static String showDialog(String prompt, String title) {
        if (DecyxConfig.SKIP_PROMPT_CONFIRMATION) {
            return prompt;
        }

        CountDownLatch latch = new CountDownLatch(1);
        PromptReviewDialog[] dialogHolder = new PromptReviewDialog[1];

        SwingUtilities.invokeLater(() -> {
            dialogHolder[0] = new PromptReviewDialog(prompt, title, latch::countDown);
        });

        try {
            latch.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        return dialogHolder[0] != null ? dialogHolder[0].getFinalPrompt() : null;
    }
}
